File: Assignment2/main.c
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MARGIN 0.2
#define MIN_ANGLE 80.0
#define MAX_ANGLE 130.0
#define MAX_NODES 1000

typedef struct {
  double x, y, z;
} Point;

typedef struct {
  int *neighbors;
  int count;
  int capacity;
} AdjList;

typedef struct {
  int *nodes;
  int length;
} Path;

int n_points;
Point points[MAX_NODES];
AdjList adj_list[MAX_NODES];
Path longest_path;
int max_length;

void add_neighbor(AdjList *list, int neighbor) {
  if (list->count >= list->capacity) {
    list->capacity = list->capacity == 0 ? 10 : list->capacity * 2;
    list->neighbors = realloc(list->neighbors, list->capacity * sizeof(int));
  }
  list->neighbors[list->count++] = neighbor;
}

double distance(Point p1, Point p2) {
  double dx = p2.x - p1.x;
  double dy = p2.y - p1.y;
  double dz = p2.z - p1.z;
  return sqrt(dx * dx + dy * dy + dz * dz);
}

double compute_angle(Point p1, Point p2, Point p3) {
  double v1x = p1.x - p2.x;
  double v1y = p1.y - p2.y;
  double v1z = p1.z - p2.z;

  double v2x = p3.x - p2.x;
  double v2y = p3.y - p2.y;
  double v2z = p3.z - p2.z;

  double dot = v1x * v2x + v1y * v2y + v1z * v2z;
  double m1 = sqrt(v1x * v1x + v1y * v1y + v1z * v1z);
  double m2 = sqrt(v2x * v2x + v2y * v2y + v2z * v2z);

  return acos(dot / (m1 * m2)) * 180.0 / M_PI;
}

bool good_angle(int *path, int path_len) {
  if (path_len < 3)
    return true;

  double angle =
      compute_angle(points[path[path_len - 3]], points[path[path_len - 2]],
                    points[path[path_len - 1]]);

  return angle > MIN_ANGLE && angle < MAX_ANGLE;
}

void dfs(int node, bool *visited, int *path, int path_len) {
  if (!good_angle(path, path_len)) {
    return;
  }

  int remaining = n_points;
  for (int i = 0; i < n_points; i++) {
    if (visited[i])
      remaining--;
  }

  if (path_len + remaining <= max_length) {
    return;
  }

  if (path_len > max_length) {
    max_length = path_len;
    memcpy(longest_path.nodes, path, path_len * sizeof(int));
    longest_path.length = path_len;
  }

  for (int i = 0; i < adj_list[node].count; i++) {
    int neighbor = adj_list[node].neighbors[i];
    if (!visited[neighbor]) {
      visited[neighbor] = true;
      path[path_len] = neighbor;
      dfs(neighbor, visited, path, path_len + 1);
      visited[neighbor] = false;
    }
  }
}

void find_longest_path() {
  bool visited[MAX_NODES];
  int path[MAX_NODES];

  for (int start = 0; start < n_points; start++) {
    memset(visited, 0, sizeof(visited));
    visited[start] = true;
    path[0] = start;
    dfs(start, visited, path, 1);

    printf("Progress: %d/%d\r", start + 1, n_points);
    fflush(stdout);
  }
  printf("\n");
}

void build_adj_list(bool **adj_matrix) {
  for (int i = 0; i < n_points; i++) {
    for (int j = 0; j < n_points; j++) {
      if (adj_matrix[i][j]) {
        add_neighbor(&adj_list[i], j);
      }
    }
  }
}

void print_chain() {
  for (int i = 0; i < longest_path.length - 1; i++) {
    printf("%d -> ", longest_path.nodes[i] + 1);
  }
  printf("%d\n", longest_path.nodes[longest_path.length - 1] + 1);
}

int main(int argc, char *argv[]) {
  if (argc < 2) {
    fprintf(stderr, "Usage: %s <input_file>\n", argv[0]);
    return 1;
  }

  FILE *f = fopen(argv[1], "r");
  if (!f) {
    perror("Error opening file");
    return 1;
  }

  n_points = 0;
  int index;
  while (fscanf(f, "%d %lf %lf %lf", &index, &points[n_points].x,
                &points[n_points].y, &points[n_points].z) == 4) {
    n_points++;
  }
  fclose(f);

  printf("Read %d points\n", n_points);

  bool **adj_matrix = malloc(n_points * sizeof(bool *));
  for (int i = 0; i < n_points; i++) {
    adj_matrix[i] = calloc(n_points, sizeof(bool));
  }

  for (int i = 0; i < n_points; i++) {
    for (int j = 0; j < n_points; j++) {
      double dist = distance(points[i], points[j]);
      double err = dist - 3.8;
      if (err >= -MARGIN && err <= MARGIN) {
        adj_matrix[i][j] = true;
      }
    }
  }

  build_adj_list(adj_matrix);

  longest_path.nodes = malloc(MAX_NODES * sizeof(int));
  max_length = 0;

  find_longest_path();

  print_chain();
  printf("Length of chain is %d\n", max_length);

  for (int i = 0; i < n_points; i++) {
    free(adj_matrix[i]);
    free(adj_list[i].neighbors);
  }
  free(adj_matrix);
  free(longest_path.nodes);

  return 0;
}

File: Assignment2/main.py
# By Simon Westlin Green
import sys
import numpy as np

# import math
from tqdm import trange

MARGIN = 0.2
MIN_ANGLE = 83
MAX_ANGLE = 127

known_angles = {}


def main():
    try:
        file = sys.argv[1]
    except IndexError as e:
        raise ValueError(
            "No input args specified, example usage: `python main.py data_q1.txt`"
        ) from e
    # the indexes in this matrix are the ones in the files - 1 because I prefer zero indexing
    matrix = read_data_points(file)
    dist = compute_distance_matrix(matrix)
    # Sub 3.8 from each cell to get them to contain error instead of distance
    err = dist - 3.8
    # print(err)
    adj_matrix = (err >= -MARGIN) & (err <= MARGIN)
    # print_bool_matrix(adj_matrix)

    adj_list = to_adj_list(adj_matrix)
    length, chain = longest_acyclic_path(adj_list, matrix)
    # print_chain(chain)
    ugly_print_chain(chain)
    # print_angles(df, chain)
    print(f"Length of chain is {length}")


# Helper functions to display information
def print_bool_matrix(matrix):
    for row in matrix[::-1]:
        for cell in row[::-1]:
            print(f"{'T' if cell else 'F'}", end="\t")
        print()


def ugly_print_chain(chain):
    for value in chain:
        print(f"{value + 1}")


def print_chain(chain):
    for value in chain[:-1]:
        print(f"{value + 1} -> ", end="")
    print(chain[-1] + 1)


def print_angles(chain, matrix):
    def gen():
        loc_chain = chain.copy()
        while len(loc_chain) >= 3:
            p1 = matrix[loc_chain.pop()]
            p2 = matrix[loc_chain[-1]]
            p3 = matrix[loc_chain[-2]]
            yield (p1, p2, p3)

    for p1, p2, p3 in gen():
        angle = compute_angle(p1, p2, p3)
        print(np.degrees(angle))


def compute_distance_matrix(matrix):
    # Vectorized version
    # I used the help of LLMs to optimize this, because I don't care enough about numpy semantics.
    # It does in essence the exact same thing as below, with some added numpy magic
    diff = matrix[:, None, :] - matrix[None, :, :]
    return np.linalg.norm(diff, axis=2)

    # dist = np.zeros((len(matrix), len(matrix)))
    # for i in range(len(matrix)):
    #     for j in range(len(matrix)):
    #         x1 = matrix[i][0]
    #         y1 = matrix[i][1]
    #         z1 = matrix[i][2]
    #         x2 = matrix[j][0]
    #         y2 = matrix[j][1]
    #         z2 = matrix[j][2]
    #         distance = math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2 + (z2 - z1) ** 2)
    #         dist[i][j] = distance
    # return dist


def read_data_points(file):
    with open(file, "r") as f:
        lines = f.readlines()
        matrix = [[]] * len(lines)
        for line in lines:
            index, x, y, z = line.split()
            matrix[int(index) - 1] = [float(x), float(y), float(z)]
    return np.array(matrix)


def to_adj_list(adj_matrix):
    n = len(adj_matrix)
    adj_list = [[] for _ in range(n)]
    for i in range(n):
        adj_list[i] = np.where(adj_matrix[i])[0].tolist()
    return adj_list


def get_chains(adj_list):
    chains = []
    visited = [False] * len(adj_list)

    for start in range(len(adj_list)):
        search_chain(start, adj_list, visited, [], chains)

    return chains


def search_chain(current, adj_list, visited, curr_chain, chains):
    visited[current] = True
    curr_chain.append(current)
    chains.append(curr_chain.copy())

    for next in adj_list[current]:
        if not visited[next]:
            search_chain(next, adj_list, visited, curr_chain, chains)

    curr_chain.pop()
    visited[current] = False


def compute_angle(p1, p2, p3):
    v1 = p1 - p2
    v2 = p3 - p2
    m1 = np.linalg.norm(v1)
    m2 = np.linalg.norm(v2)
    return np.arccos(np.dot(v1, v2) / (m1 * m2))


def good_angle(path, matrix):
    global known_angles
    if len(path) < 3:
        return True
    key = (path[-3], path[-2], path[-1])
    angle = known_angles.get(key)
    if angle is not None:
        return angle > MIN_ANGLE and angle < MAX_ANGLE
    angle = np.degrees(
        compute_angle(matrix[path[-3]], matrix[path[-2]], matrix[path[-1]])
    )
    known_angles[key] = angle
    return angle > MIN_ANGLE and angle < MAX_ANGLE


def longest_acyclic_path(graph, matrix):
    max_length = 0
    longest_path = []

    def dfs(node, visited, path):
        nonlocal max_length, longest_path

        if not good_angle(path, matrix):
            return

        if len(path) > max_length:
            max_length = len(path)
            longest_path = path.copy()

        remaining = len(graph) - len(visited)
        if len(path) + remaining <= max_length:
            return

        for neighbor in graph[node]:
            if neighbor not in visited:
                visited.add(neighbor)
                path.append(neighbor)
                dfs(neighbor, visited, path)
                path.pop()
                visited.remove(neighbor)

    # Try starting from each vertex
    for start in trange(len(graph)):
        visited = {start}
        dfs(start, visited, [start])

    return max_length, longest_path


if __name__ == "__main__":
    main()
